spec/ordinary-and-exotic-object-behaviours.html

```diff
@@ -0,0 +1,316 @@
+        <emu-alg>
+          1. Return ! SetImmutablePrototype(_T, _V_).
+        </emu-alg>
+      </emu-clause>
+      
+      <emu-clause id="sec-tuple-exotic-objects-isextensible">
+        <h1>[[IsExtensible]] ()</h1>
+        <p>When the [[IsExtensible]] internal method of a Tuple exotic object _T_ is called, the following steps are taken:</p>
+        <emu-alg>
+          1. Return *false*.
+        </emu-alg>
+      </emu-clause>
+
+      <emu-clause id="sec-tuple-exotic-objects-getownproperty-p">
+        <h1>[[GetOwnProperty]] ( _P_ )</h1>
+        <p>When the [[GetOwnProperty]] internal method of a Tuple exotic object _T_ is called with property key _P_, the following steps are taken:</p>
+        <emu-alg>
+          1. Assert: IsPropertyKey(_P_) is *true*.
+          1. If Type(_P_) is String, then
+            1. Let _numericIndex_ be ! CanonicalNumericIndexString(_P_).
+            1. If _numericIndex_ is not *undefined*, then
+              1. Let _value_ be ? TupleGet(_T_, _numericIndex_)
+              1. If _value_ is *undefined*, return *undefined*.
+              1. Return the PropertyDescriptor { [[Value]]: _value_, [[Writable]]: *false*, [[Enumerable]]: *true*, [[Configurable]]: *false* }.
+          1. Return *undefined*
+        </emu-alg>
+      </emu-clause>
+
+      <emu-clause id="sec-tuple-exotic-objects-hasproperty-p">
+        <h1>[[HasProperty]] ( _P_ )</h1>
+        <p>When the [[HasProperty]] internal method of a Tuple exotic object _T_ is called with property key _P_, the following steps are taken:</p>
+        <emu-alg>
+          1. Assert: IsPropertyKey(_P_) is *true*.
+          1. If Type(_P_) is String, then
+            1. Let _numericIndex_ be ! CanonicalNumericIndexString(_P_).
+            1. If _numericIndex_ is not *undefined*, then
+              1. Return ? IsValidTupleIndex(_T_, _numericIndex_)
+          1. Return *false*
+        </emu-alg>
+      </emu-clause>
+
+      <emu-clause id="sec-tuple-exotic-objects-defineownproperty-p-desc">
+        <h1>[[DefineOwnProperty]] ( _P_, _Desc_ )</h1>
+        <p>When the [[DefineOwnProperty]] internal method of a Tuple exotic object _T_ is called with property key _P_, and Property Descriptor _Desc_ the following steps are taken:</p>
+        <emu-alg>
+          1. Assert: IsPropertyKey(_P_) is *true*.
+          1. If Type(_P_) is String, then
+            1. Let _numericIndex_ be ! CanonicalNumericIndexString(_P_).
+            1. If _numericIndex_ is not *undefined*, then
+              1. Let _current_ be ? TupleGet(_T_, _numericIndex_)
+              1. If _current_ is *undefined* return *false*.
+              1. If IsAccessorDescriptor(_Desc_) is *true*, return *false*.
+              1. If _Desc_.[[Writable]] is present and has value *true*, return *false*.
+              1. If _Desc_.[[Enumerable]] is present and has value *false*, return *false*.
+              1. If _Desc_.[[Configurable]] is present and has value *true*, return *false*.
```

Comment:
Editorial suggestion: These four conditionals are to be checked regardless of what TupleGet outputs, so I think it makes sense to put them before the TupleGet call.